### DISCUSSION TITLE
About ambiguous implicit instances in Cats module

### DISCUSSION THREAD

**Discussion by Iltotore** [Iltotore/iron#95] [Q&A] [answered]
Summoning again @gvolpe because I forgot to ask it when the PR got merged: why do we need to make our given `Hash` low priority?

Is this because it can conflict with `Eq` since `Hash <: Eq`? If yes, then shouldn't the same problem occur with `Ord` and `Eq`, `Group` and `Monoid` etc... ?

**Comment by gvolpe** [comment, marked as answer]
The first reason is the clash with the `Eq` instance, that's correct. But I think it has to do with how it's defined:

```scala
trait Hash[@sp A] extends Any with Eq[A] with Serializable
```

Versus

```scala
trait Order[@sp A] extends Any with PartialOrder[A]

trait PartialOrder[@sp A] extends Any with Eq[A] 

trait Eq[@sp A] extends Any with Serializable
```

`Hash` extends `Serializable` directly, and that seems to affect implicit resolution, but that's just a wild guess. You can see how `Eq` also extends `Serializable`, but in the eyes of both `Order` and `PartialOrder` it's another layer down in the mixin.

 I only know that the low-priority pattern is quite common for typeclass instances, so I just moved this one there and everything else compiled.

**Reply by Iltotore** [reply]
> but in the eyes of both Order and PartialOrder it's another layer down in the mixin.

So there is no conflict between `Eq` and `Order`/`PartialOrder`? That's my actual question to know if these instances should also be moved or not.

**Reply by gvolpe** [reply]
There are no conflicts at all. It's all checked in the [tests](https://github.com/Iltotore/iron/blob/dev/v2.0/cats/test/src/io/github/iltotore/iron/CatsSuite.scala#L36) at compile time.
